src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -2,68 +2,113 @@
 import java.util.function.Function;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 public class UserServiceImpl implements UserService {
 
     @Override
     public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(User::getFirstName)
+                .sorted(Comparator.reverseOrder())
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .sorted(Comparator.comparing(User::getAge)
+                        .reversed()
+                        .thenComparing(User::getFirstName))
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .distinct()
+                .flatMap(user -> user.getPrivileges().stream())
+                .collect(Collectors.toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .filter(user -> user.getPrivileges().contains(Privilege.UPDATE))
+                .filter(user -> user.getAge() > age)
+                .findFirst();
     }
 
     @Override
     public Map<Integer, List<User>> groupByCountOfPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .collect(Collectors
+                        .groupingBy(user -> user.getPrivileges().size()));
     }
 
     @Override
     public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .mapToInt(User::getAge)
+                .average()
+                .orElse(-1d);
     }
 
     @Override
     public Optional<String> getMostFrequentLastName(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        Map<String, Long> lastNameFreq = users.stream()
+                .map(User::getLastName)
+                .collect(Collectors.
+                        groupingBy(Function.identity(), Collectors.counting()));
+        Long max = lastNameFreq.values().stream().max(Long::compareTo).orElse(0L);
+        return lastNameFreq.values().stream()
+                .filter(value -> value.equals(max))
+                .count() != 1 ? Optional.empty() : lastNameFreq.entrySet().stream()
+                .max(Map.Entry.comparingByValue()).map(Map.Entry::getKey);
+
```

Comment:
An alternative solution, could be this:
 return users.stream()
                .collect(groupingBy(User::getLastName, counting()))
                .entrySet().stream().
                        collect(groupingBy(Map.Entry::getValue))
                .entrySet().stream()
                .max(Map.Entry.comparingByKey())
                .map(Map.Entry::getValue)
                .filter(v -> v.size() == 1)
                .map(v -> v.get(0).getKey());

Try to understand how it works and you can use it if you like it more

Reply:
Fixed and understand.